src/lib.rs

```diff
@@ -3,30 +3,45 @@ use kvmi_sys::{kvmi_qemu2introspector, kvmi_introspector2qemu};
 use std::ffi::{CString};
 use std::ptr::{null_mut};
 use std::os::raw::{c_void, c_uchar, c_int};
+use std::sync::{Mutex, Condvar};
 
 
 #[derive(Debug)]
 pub struct KVMi {
     ctx: *mut c_void,
+    dom: *mut c_void,
+}
+
+#[derive(Debug)]
+struct KVMiCon {
+    dom: *mut c_void,
+    guard: Mutex<bool>,
+    condvar: Condvar,
 }
 
 extern "C" fn new_guest_cb(dom: *mut c_void,
                            uuid: *mut [c_uchar; 16usize],
-                           ctx: *mut c_void) -> c_int {
+                           cb_ctx: *mut c_void) -> c_int {
     println!("new guest cb !");
+    let kvmi_con = unsafe { &mut *(cb_ctx as *mut KVMiCon) };
```

Comment:
If the `KVMiCon` structure is passed as opaque to C, it should not be required

However, this function should be `unsafe extern "C" fn`, since you are de-referencing a pointer. BTW, you should test this pointer (could be `std::ptr::null_mut()`)

Reply:
Hi @chifflier, @Wenzel 

Correct me if I was wrong. Because `cb_ctx` will be used as a `C` pointer to a `C` struct on the other side of FFI (I suppose that there is a corresponding `C` struct), so if `KVMiCon` has no `#[repr(C)]` then there is no warranty that two sides have an unique memory layout.